### ISSUE TITLE
CLI: Cannot bulk-export MonitoringAlertPolicy

### ISSUE THREAD

**Issue by tsawada** [GoogleCloudPlatform/k8s-config-connector#589]
## Bug Description

I cannot export any AlertPolicy rule in my project using the cli with the following error:

```
$ config-connector bulk-export --project=my-project
error getting next YAML: error getting unstructured: error getting next unstruct: error converting asset '//monitoring.googleapis.com/projects/***/policies/***' with kind 'monitoring.googleapis.com/AlertPolicy' to skeleton: error importing resource name to TF state: error importing resource: project: required field is not set
```

### Config Connector Version

```
$ config-connector version
1.70.0
```

Let me know if there's anything else I can provide. Thanks!

**Comment by diviner524** [comment]
@tsawada Thank you for reporting this issue! 

While we are investigating, could you please help to provide some more information:

1. The error message in your example is:  

    error converting asset '//monitoring.googleapis.com/projects/***/policies/***'

    While monitoring alert policy should have a resource name like '//monitoring.googleapis.com/projects/***/alertPolicies/***'

    Could you please confirm if it is an edit error?

2. If possible, could you please also provide one example of the alertpolicy rule from gcloud output, something like below:

$ gcloud alpha monitoring policies describe <yourpolicyname>

**Comment by tsawada** [comment]
Hi @diviner524 Thanks for your quick response!

1.
I did not edit that part of the log. The resource name in the log is actually `/policies/` not `/alertPolicies/`.
I can also see the detail of the alert policy in the Console on `https://console.cloud.google.com/monitoring/alerting/policies/<second *** in the log>?project=my-project` so I believe those are still the alertPolicies.

2.
Here. I edited only the part annotated ${like_this} and didn't remove anything else.
```
$ gcloud alpha monitoring policies describe projects/${my-project}/alertPolicies/${policy_id}
alertStrategy: {}
combiner: OR
conditions:
- conditionThreshold:
    aggregations:
    - alignmentPeriod: 300s
      perSeriesAligner: ALIGN_MEAN
    comparison: COMPARISON_GT
    duration: 120s
    filter: metric.type="cloudsql.googleapis.com/database/cpu/utilization" resource.type="cloudsql_database"
    thresholdValue: 0.5
    trigger:
      count: 1
  displayName: Cloud SQL Database - CPU utilization
  name: projects/${my-project}/alertPolicies/${policy_id}/conditions/14459391614995787414
creationRecord:
  mutateTime: '2021-03-21T08:15:07.463908420Z'
  mutatedBy: ${email}
displayName: Cloud SQL Database - CPU utilization
enabled: true
mutationRecord:
  mutateTime: '2021-03-24T00:42:53.300196246Z'
  mutatedBy:${email}
name: projects/omega-fabric-148305/alertPolicies/${policy_id}
notificationChannels:
- projects/omega-fabric-148305/notificationChannels/${notify_id1}
- projects/omega-fabric-148305/notificationChannels/${notify_id2}
```

**Comment by diviner524** [comment]
@tsawada Sorry for the delayed response. We are still working with Cloud Asset Inventory team to investigate this issue. Will post here when we have any new updates.

**Comment by diviner524** [comment]
@tsawada We have identified the issue and will be working on a fix.

At the same time, could you please let us know if this issue is still a blocker for you?

Note that, if your goal is to export/import all resources under that project (including MonitoringAlertPolicy resources) then this issue could be a blocker which will prevent you from directly running  below command.

```
$ config-connector bulk-export --project=my-project
```

However if you don't require MonitoringAlertPolicy resources to be exported, there is one workaround.

- You can first try [Cloud Asset Inventory Export](https://cloud.google.com/asset-inventory/docs/exporting-to-cloud-storage#exporting_all_resource_names_without_metadata), which will export resources under a project as a json file, in a Storage Bucket.

- Then you can run command like below to skip exporting MonitoringAlertPolicy resources.  (Assuming your json file is called "export.json".

```
$ cat export.json | jq '. | select( .asset_type != "monitoring.googleapis.com/AlertPolicy" )' | config-connector bulk-export
```

It is also explained in our doc [here](https://cloud.google.com/config-connector/docs/how-to/import-export/bulk-export#exporting_your_project).

**Comment by tsawada** [comment]
Thanks @diviner524 . I ran `config-connector` with `--on-error=continue` so other type of resources are fine. However I'm interested in AlertPolicy exports so that we can use it as a basis for our configuration. I'm looking forward for the fix 😁

**Comment by diviner524** [comment]
@tsawada Thanks for the reply! Glad to hear that you have found a workaround. We will update this thread when the fix is ready.

**Comment by ckiosidis** [comment]
Hey @diviner524  this issue affects exporting single alertpolicy resources as well, example:

`config-connector export //monitoring.googleapis.com/projects/gcpProject/alertPolicies/policyId`

Is there an ETA for the fix?

**Comment by diviner524** [comment]
@ckiosidis Yes, this is a bug on alertpolicy resource so exporting it alone will also run into the same issue.

Sorry we don't have an ETA at this moment.

Could you please let us know the impact of this issue? Is it currently a blocker for you? This will help us to better prioritize our tasks.

**Comment by ckiosidis** [comment]
Hey @diviner524 
We are moving our resources to yamls and I guess we cannot do that for Monitoring Alert Policies.
It's not critical per se, since the policy is working as it is, so we won't complain much if we have to wait.

**Comment by tsawada** [comment]
With the latest 1.81.0, the error message has slightly changed.
```
error getting next YAML: error getting unstructured: error getting next unstruct: error getting '//monitoring.googleapis.com/projects/<numbers>/policies/<numbers>': error fetching live state: error getting ID for resource: error getting server-generated resource ID from the value of 'status.name': error extracting out the value segment from the resolved template 'projects/<numbers>/policies/<numbers>' using template '^projects/[^/]+/alertPolicies/([^/]+)$'
```

I'd appreciate if you could provide some ETA. This is blocking us from using alertPolicies.

**Comment by diviner524** [comment]
@tsawada Sorry for the inactivity, the team is moving slowly on bug fixes due to capacity constraints. I have added your comments in the internal tracking ticket and mentioned this is now a blocker for a customer. We will prioritize it and see if the work will fit into our upcoming sprints.

**Comment by mikesoule** [comment]
@diviner524 I have just under 100 policies I'll have to create in Terraform manually without this. Is there an ETA?

**Comment by caieo** [comment]
@mikesoule, We are prioritizing a fix in our coming sprints and unfortunately don't have a solid ETA for you. Sorry for the inconvenience.

**Comment by tsawada** [comment]
I opened a PR to fix exporting single MonitoringAlertPolicy: #659 
For bulk-export, I think there's something wrong in the Asset Inventory. https://issuetracker.google.com/issues/232441106
